Community Infra connections for staging

cori opened the pull request:
## Links
* [README](ci/README.md)
* [diagram](https://docs.google.com/drawings/d/1KmJhEIkqhS2VBxhiZNYJSiO0iUX8IrVQW4soiHcE_0k/edit)

## Changes:
* updated `.circleci/config.yml` to add deploy steps
* added scripts in `ci` to handle CircleCI-based and local-to-the-community-hosts deploy scripts

## How To Test:
* merge changes into staging
* watch CircleCI job execute (and, hopefully, succeed!)
* check staging.glitch.com for update

## Feedback I'm looking for:
* is this understandable?
* is it clear how development will change?
* what additional docs or guidance is needed?
* is it stable?
* what else do we need to address (aside from moving into prod)?

## Things left to do before deploying:
* i think this can be merged and we can start testing it - it shouldn't affect anyone not merging to `staging`, which is new, so....

clottman commented on .circleci/config.yml:
```diff
@@ -3,75 +3,140 @@
-          command: "sudo npm install -g pnpm@3.x"
+          name: setup-ssh
+          command: cp ./ci/ssh.config ~/.ssh/config
+      # - run:
```
Any reason we still need this commented out? Or should we just axe it?

cori replied:
i removed this one. There were others in some of the testing jobs that I don't recall modifying (although I probably did) so I reset them - those were outside my scope.

keithk commented:
Haven't had a chance to pour through this yet, but wondering if when we're ready to deploy the comments (especially to notes to self) will be removed, or if there are any of them that remain questions that we should answer?

clottman commented on ci/README.md:
```diff
@@ -0,0 +1,59 @@
+
+## Overall Process
+
+1. local dev &#x2672;
```
What's this character?

cori replied:
This and the list numbering were intended to be viewed in the rendered file, but nevertheless they weren't adding anything so I removed them.

clottman commented on ci/README.md:
```diff
@@ -0,0 +1,59 @@
+## Overall Process
+
+1. local dev &#x2672;
+1. push to `staging`
```
Can you clarify what this means? Like is the expected flow that we merge a pull request to staging? Or push from the command line?

cori replied:
I think there is some discussion to be had on this front right now, and I'm more or less agnostic about it. Either would work from the technical side of things, as would skipping staging entirely (once we've got master wired up)

keithk replied:
I'm thinking that we open PRs to staging, verify it on staging.glitch.com (because it's on the same infrastructure as the production site, different than testing locally), and then git merge staging into master

clottman commented on ci/README.md:
```diff
@@ -0,0 +1,59 @@
+1. local dev &#x2672;
+1. push to `staging`
+   1. CircleCI picks up the commit
+      1. lints, builds, runs tests, just like it does right now
```
This is minor but can we make this an actual numbered list? So when discussing instructions we can say ie, "I just followed step 3 and then this happened"

cori replied:
done

clottman commented on ci/README.md:
```diff
@@ -0,0 +1,59 @@
+1. push to `staging`
+   1. CircleCI picks up the commit
+      1. lints, builds, runs tests, just like it does right now
+      1. checks if the asset is in s3 [remotely executed on a worker]
```
When you say "checks if the asset is.." - what asset are you talking about?

cori replied:
updated, but let me know if still unclear

clottman commented on ci/README.md:
```diff
@@ -0,0 +1,59 @@
+      1. sends `deploy` command, which gets the asset from S3 and deploys it to the local box [remotely executed on each worker]
+1. merge to `master`
+   1. CircleCI picks up the commit
+      1. same process as for staging @ 2.1.1 above, but deploying to the production boxes
```
Same comment here re: numbering the list. Also an example of why - which "2.1.1" is this referring to?

cori replied:
see prev.

clottman commented on ci/README.md:
```diff
@@ -0,0 +1,59 @@
+
+### No secrets in the Glitch-Community repo
+
+A central tenet to the community CI work is that the community repo shouldn't house any secrets (even encrypted ones). This means that any information we need to get from somewhere else (AWS, primarily) needs to be proxied through something that has the appropriate permissions. To avoid additional complexity, I just relied on the built-in access that the community host machines already have by vi ...
```
I don't understand what this means. In the past if I needed to add a new .env value (maybe an API key to reach Prismic or a new service), I'd add it to .env. What should I do in the new system if I need to add a key? Where do I find the current key, for example if I need to cycle an API key?

cori replied:
As the new deploy is currently structured, the `.env` file is stored in the s3 bucket and would need to be updated there, which (again as things stand on 2019-12-12) will require coordination with someone with access - potentially via an Infrastructure story in ClubHouse, but TBD).

In going that route I looked at the current `.env` file and its history and it seemed like it changed only rarely so this seemed like a fairly small burden. There's only one key in there right now that seemed like it should be protected (which is what this configuration is attempting to retain) but it seemed likely that number was only going to go up. I'm open to discussions about whether this is the correct, or even a workable approach.

clottman commented on ci/README.md:
```diff
@@ -0,0 +1,59 @@
+### Things I chose _not_ to do
+
+* prevent overlapping CircleCI deploys
+  * there are so few boxes, deploys should happen so uickly, and the number of simulatneous deploys is so low this didn't seem worth any effort right now
```
Minor - typo here on 'uickly'

clottman commented on ci/README.md:
```diff
@@ -0,0 +1,59 @@
+* prevent overlapping CircleCI deploys
+  * there are so few boxes, deploys should happen so uickly, and the number of simulatneous deploys is so low this didn't seem worth any effort right now
+* implement `no-deploy` or `no-test` flags
+  8 again this didn't seem needed right away; we can implement them later if we care to
```
8?

cori replied:
*

clottman commented on ci/check-deploy-source.sh:
```diff
@@ -0,0 +1,27 @@
+#####
+
+# check req params - we need the env and a sha to use for file manipulation
+if [ 2 -ne "$#" ]; then
```
I'm no shell wizard and this is hard to google - what does `$#` mean? At what point would it be equal to 2?

cori replied:
`$#` is the number of arguments passed in to the script, but you're right, bash is very hard to google. Hopefully the updated comment is clearer.

clottman commented on ci/check-deploy-source.sh:
```diff
@@ -0,0 +1,27 @@
+source /opt/glitch-community/ci/env
+
+# check S3 for the asset; tell caller the result
+# aws s3 ls s3://community-bootstrap-bucket20191205165831056600000001 | grep "$CIRCLE_SHA"; code=$?
```
what's the commented out part doing here? Do we need it?

cori replied:
that was a left-over - gone now

clottman commented on ci/local-deploy.sh:
```diff
@@ -0,0 +1,47 @@
+
+source ci/env
+
+# we run npm i here in case pm2 is not available but we could probably just swallow that error
```
What does "in case pm2 is not available" mean? What conditions would cause that, and how do we recover?

cori replied:
hopefully more clear now

clottman commented on ci/local-deploy.sh:
```diff
@@ -0,0 +1,47 @@
+# we run npm i here in case pm2 is not available but we could probably just swallow that error
+npm i && npm run stop && wait
+
+# avoid cruft like deleted files from hanging around; currently removed folders will still persist
```
minor - the 'from' in the first part of the sentence confuses me a bit. Maybe it's a typo?

cori replied:
clearer?

clottman commented on ci/local-deploy.sh:
```diff
@@ -0,0 +1,47 @@
+#   install the deps, run the app
+npm i && wait
+
+#   one can temporarily add a pseudo-random pause here
```
When would this be necessary?

cori replied:
I used this extensively in testing long-running deploys, and in case we need to run some more tests of that nature I didn't want to remove it yet.

clottman commented on package.json:
```diff
@@ -26,8 +26,9 @@
     "lint:src": "eslint --config src/.eslintrc.js src",
-    "prestart": "npm run watch:lint:server & \n if if-env BUILD_TYPE=watcher; then \n npm run watch:webpack & \n fi",
-    "start": "pm2-runtime start ecosystem.config.js",
+    "prestart": "if if-env BUILD_TYPE=memory; then \n npm run watch:lint:server & \n fi & \n if if-env BUILD_TYPE=watcher; then \n npm run watch:webpack \n npm run watch:lint:server & \n fi",
```
We probably want to clean this up and remove BUILD_TYPE=watcher - is that on your todo list @cori or something we should make a ticket for? 

Also - what's this change doing?

cori replied:
I had to make sure that we did _not_ run the lint watcher in "serving" ("static") mode (in either staging or production) because the CircleCI job would never exit. This may not be the right or best way to handle this, and any other BUILD_TYPE changes are outside my remit - if there's a better change that would accomplish what I was looking for I'm happy to hear it!

GregWeil replied:
This changes the default behavior when `BUILD_TYPE` isn't set. It's supposed to act like `memory` but this won't run the server linter.

I would like to keep BUILD_TYPE=watcher around, it doesn't restart webpack on server restart which is helpful when making server changes.

cori replied:
I can try to replicate that logic - is there somewhere I should be looking for what the expected behavior is so I can try to match it? Is it basically just that unset `BUILD_TYPE` should behave like `memory`?

GregWeil replied:
The readme has the default value for each var, but it doesn't go that in depth on their effects. Unset BUILD_TYPE should be equivalent to memory.

cori replied:
I believe the most recent commit handles the unset BUILD_TYPE appropriately, or it seemed to in my local testing.

cori replied:
This is now working as expected, I _think_, but we may want to extract this complicated command into a tiny script. Deferring that for the moment until we talk about https://github.com/FogCreek/Glitch-Community/pull/1121#discussion_r359072657

cori commented on .circleci/config.yml:
```diff
@@ -3,75 +3,140 @@
           keys:
-            - deps-{{ checksum "shrinkwrap.yaml" }}
+            # in Glitch we key on .Branch, but this is probably just as good
+            # diff branches will probably have different checksums
```
that was an out-of-date comment, sorry!

cori commented:
> Haven't had a chance to pour through this yet, but wondering if when we're ready to deploy the comments (especially to notes to self) will be removed, or if there are any of them that remain questions that we should answer?
> 
> There also appears to be a good amount of code that is entirely commented out, so once we're sure that it's not required can we remove it (so it's not dead code)

I mostly cleaned up comments, but I didn't get back to `.circleci/config.yml` before I sent the PR, so that one didn't get cleaned up. The subsequent commit should address much of that, so I don't think there should be much a) dead code or b) overly-personal comments left, but let me know what you find!

mads-hartmann commented on ci/README.md:
```diff
@@ -0,0 +1,59 @@
+  |
+  - (these files all run on the hosts in the Glitch Infrastructure)
+  - check-deploy-source.sh - checks to see if the asset for the current sha is in S3
+  - upload-asset.sh - handles uploading the provided asset to the appropriate bucket and marking the correct sha
```
Would it be possible to remove `check-deploy-source.sh` entirely and instead invoke `upload-asset.sh` in `config.yaml` if you generated AWS access key/secret that only allowed reading/writing to the community bucket and added it to the CirleCI environment?

If that's feasible I think that would make the deploy.sh script easier to read :)

cori replied:
We could definitely do that; one of my unifying principles was not to have any secrets in the Community configuration, but we could achieve this in CircleCI itself without secrets in the _repo_, that's just not something I see us doing elsewhere right now, so I didn't want to open up a new layer of dependencies or a new pattern if I could avoid it, but maybe I bent over too far?

I guess it's a matter of preference and priority. I can implement that change without too much difficulty, I think, if we agree that it would be a substantial improvement - it would make the script easier to read and probably a _bit_ less fragile.

mads-hartmann replied:
I think we upload our bootstrap scripts directly from CIrcleCI and the editor does it for their assets I think. For example we have both  `STAGING_AWS_BOOTSTRAP_SECRET` and `PRODUCTION_AWS_BOOTSTRAP_SECRET` in our circle environment variables right now :)

cori replied:
Yep! Those are encrypted and stored in the repo at `/ci/secrets/secure.env`. which was what I was working to avoid doing in the Glitch-Community repo.

We could use CircleCI builtins to set up AWS access and manage the secrets in CircleCI, which was the alternative I was referring to (that I don't think we do anywhere else)

mads-hartmann replied:
The are also set in the CircleCI environments variable config :) https://circleci.com/gh/FogCreek/Glitch/edit#env-vars

mads-hartmann replied:
I have a feeling they don't actually need to be in `/ci/secrets/secure.env` but I haven't verified this

cori replied:
yup i totally missed those - I saw them being set via the top of `deploy.sh`, but this might make it more explicit; thanks!

eestolano replied:
I just verified it myself.

There are a couple of scripts (like `hostnames-by-role`) that get secrets from the CircleCI env if they're run in CircleCI, but get the secrets from `/ci/secrets/secure.env` when run outside of CircleCI. CircleCI itself doesn't get the key for the repo.

eestolano replied:
And, zooming out, I agree with Mads' suggestion to get rid of `check-deploy-source.sh` and run `upload-asset.sh` directly from CircleCI itself. It would be much simpler.

keithk replied:
Just checking on the status of these suggestions, and if we're moving forward w/ them @cori ?

cori replied:
Yep I'm still working on these changes. hope to have them out by EoD today.

keithk commented on ci/deploy.sh:
```diff
@@ -0,0 +1,72 @@
+  # shoving the asset to the destination server and brute forcing the deploy
+  # this is super-naive, and we mightr want to handle errors more explicitly
+  catch() {
+    echo "something bad happened; let us see if just pushing the asset will work"
```
What happens if this doesn't work?

cori replied:
That represents a serious problem that probably needs investigation. Probably the easiest way forward to fix an ongoing deploy would be to terminate the host, but if it happens often we'd want to see if something in the process is broken.

cori replied:
Let me know if we need to discuss this further, @keithk!

mads-hartmann commented on ci/env:
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env bash
+
+if [ -n "${CIRCLE_BRANCH:-}" ]
```
You could collapse all of this into a single switch statement:

```sh
case "${CIRCLE_BRANCH}" in 
    master)
        export ENVIRONMENT=production
        export BOOTSTRAP_BUCKET=
        export JUMP_DOMAIN=jump.glitch.com
        ;;
    staging)
        export ENVIRONMENT=staging
        export BOOTSTRAP_BUCKET=community-bootstrap-bucket20191205165831056600000001
        export JUMP_DOMAIN=jump.staging.glitch.com
        ;;
    *)
        echo "Unknown branch '$CIRCLE_BRANCH'"
        exit 1
esac
```

cori replied:
in some places we know the `ENVIRONMENT` (because we set it) but not the `CIRCLE_BRANCH` (because that's only available inside `deploy.sh` which runs on the executor). There's probably a better way, though.

mads-hartmann replied:
Ah! I didn't catch that. If you could express that with a comment in the script that would be helpful :) 

I personally prefer switches of if's but that's just a personal preference so feel free to ignore this version:

```sh
# If the ENVIRONMENT isn't known we'll infer it from the branch
if [[ -z "${ENVIRONMENT}" ]]
then
case "${CIRCLE_BRANCH}" in 
    master)
        export ENVIRONMENT=production
        ;;
    staging)
        export ENVIRONMENT=staging
        ;;
    *)
        echo "Can't infer ENVIRONMNET from branch '$CIRCLE_BRANCH'"
        exit 1
esac
fi

# Set the right BOOTSTRAP_BUCKET and JUMP_DOMAIN based on the active ENVIRONMENT
case "${ENVIRONMENT}" in 
    production)
        export BOOTSTRAP_BUCKET=
        export JUMP_DOMAIN=jump.glitch.com
        ;;
    staging)
        export BOOTSTRAP_BUCKET=community-bootstrap-bucket20191205165831056600000001
        export JUMP_DOMAIN=jump.staging.glitch.com
        ;;
    *)
        echo "Unknown environmnet '$ENVIRONMENT'"
        exit 1
esac
```

cori replied:
Updated!

mads-hartmann commented on ci/local-deploy.sh:
```diff
@@ -0,0 +1,48 @@
+#####
+# THIS SCRIPT RUNS ON ALL COMMUNITY WORKERS
+# It manages the running site, gets the requested asset, and deploys and installs it. 
+# This script suffers from a bootstrap problem - if you change this script 
```
I think the way we get around the bootstrap problem in our `backend-local.sh` scripts is that the `parallel` command will transfer the script to the host before executing it.

cori replied:
ah using `basefile` and `cleanup`. That's a good tip. I'm not using `parallel` anywhere here since we're explicitly counting on hosts being deployed to one by one, but if this bootstrap problem becomes onerous it might be worth considering (or something similar)

eestolano replied:
I'd advise solving the bootstrap problem now; it's likely to cause problems at an extremely inconvenient time. I think the easiest way is to `scp` this script over into `/tmp`, and run that copy of it. I don't think it's important to remove it afterwards.

(If you want to use `parallel` with `basefile` and `cleanup`, you could set a parallelism level of 1.)

cori replied:
This is addressed in the latest set of commits.

eestolano commented on ci/deploy.sh:
```diff
@@ -0,0 +1,72 @@
+
+  #check if the asset is already in S3
+  ssh -o "ProxyJump $JUMP_DOMAIN" -o StrictHostKeyChecking=no "$name.$ENVIRONMENT" "bash --login -c \"cd /opt/glitch-community && ci/check-deploy-source.sh $ENVIRONMENT $CIRCLE_SHA\"" > /dev/null 2>&1; code=$?
+  S3_LOOKUP_RESULT="$code"
```
This may be a moot point if you end up getting rid of `check-deploy-source` entirely, but you could simplify a couple of things here:

* `-e` doesn't exit on errors for commands in `if` statements.
* You already have CircleCI deploying ~/.ssh/config in advance; with a suitable SSH config file, you won't need those SSH options in the command line.
* I could be wrong, but it _looks_ like `check-deploy-source` cares neither what directory it's in nor whether  you're using a login session

So you could do something like the following, and also set `-e`:

```
if ssh "$name.$ENVIRONMENT" "/opt/glitch-community/ci/check-deploy-source.sh $ENVIRONMENT $CIRCLE_SHA" > /dev/null 2>&1; then
   # &c &c &c
fi
```

Or, if `-e` doesn't work for more complicated reasons, you can always `set +e` immediately before the problematic commands, and `set -e` again afterwards.

cori replied:
I do think this is moot with recent changes.

eestolano commented on ci/deploy.sh:
```diff
@@ -0,0 +1,72 @@
+# * connect env and branch to remove hard-coded vals
+# * should this connect to Honeycomb? I think it would be valuable.
+
+# first get the list of hostnames - we could do this on any host, but we know the worker has the code
```
You can run `hostnames-by-role` directly from the CircleCI executor, if you set `AWS_DESCRIBE_INSTANCES_KEY` and `AWS_DESCRIBE_INSTANCES_SECRET` appropriately in the CircleCI env. That's how we do it for the `Glitch` repo deploys.

cori replied:
I suppose having that call here would save an http call and associated latency and point of failure. You think that's worth the overhead of keeping a separate copy of that script here (i don't have the Glitch repo available here so I'd need to copy the script if I understand what you mean).

eestolano replied:
Oh, right, you'd have to copy the script. I missed that somehow.

Avoiding the single point of failure is worthwhile, but you can do that without copying the script: use the jump box fleet instead of the worker box (`jump.staging.glitch.com` for staging and `jump.glitch.com` for prod)

(Also, the ssh command can be a bit simpler; the `hostnames-by-role` command doesn't care about having a bash login session or being in any particular directory.)

cori replied:
I cleaned these up a little; I'm going to resolve this and you can take another look if you care to.

eestolano commented on package.json:
```diff
@@ -26,8 +26,9 @@
     "lint:src": "eslint --config src/.eslintrc.js src",
-    "prestart": "npm run watch:lint:server & \n if if-env BUILD_TYPE=watcher; then \n npm run watch:webpack & \n fi",
-    "start": "pm2-runtime start ecosystem.config.js",
+    "prestart": "if if-env BUILD_TYPE=\"\"; then \n npm run watch:lint:server & \n fi & \n if if-env BUILD_TYPE=memory; then \n npm run watch:lint:server & \n fi & \n if if-env BUILD_TYPE=watcher; then \n npm run watch:webpack \n npm run watch:lint:server & \n fi",
```
Maybe the prestart could be a tiny script? This one-liner is hard to read.

cori replied:
I think this might be a useful change. In the long run `package.json` is something the community team will manage, so I'll defer to you on this @keithk.

GregWeil commented on src/components/collection/collection-result-item.styl:
```diff
@@ -11,9 +11,3 @@
-  > div
-    background-color: private
-  > div:hover
-      background-color: general-link
```
This doesn't seem like an intentional part of your pr, but this code was already removed from ~community. Maybe you got a bad merge/git doesn't see this branch as merged?

cori replied:
Oh I forgot to respond to this but Mads asked about "all the html file changes" as well this morning.

I'm pretty sure this is just related to the fact that the staging branch is new and I've only intermittently been merging changes from `Glitch-Community/master`. I think when this is merged into `staging` and we merge `master` in again it'll be easier to make sure everything looks as it should. But if it's concerning I can dig deeper first.

GregWeil replied:
I think if master gets pulled into staging these changes will stop showing up without having to pull anything into this pr.

cori commented:
All right! After a mess of commits I think I've addressed the central concerns / suggestions here. We are now:

* uploading assets to s3 directly from CircleCI
* avoiding problems with updating local-deploy.sh by sending it to the host's `/tmp` directory and executing from there

that's _most_ of what I addressed here, along with a few other minor changes.

cori commented:
@clottman I'm not sure where you requested changes and whether or not they've been addressed, but maybe you can take another look? I couldn't re-request a review from you.

clottman commented on ci/README.md:
```diff
@@ -0,0 +1,59 @@
+2. push to `staging`
+   i. CircleCI picks up the commit
+      a. lints, builds, runs tests, just like it does right now
+      b. queries s3 to determine if the build archive is already present [remotely executed on a worker]
```
Given the changes you mentioned in one of your last comments on this PR, is this doc line still accurate?

cori replied:
Nope, that’s a good catch! I plan on updating the readme and associated diagram,  but figured I’d wait to see if more changes were needed; should have made that clear.

eestolano commented on ci/deploy.sh:
```diff
@@ -0,0 +1,48 @@
+if [ ${code} -ne 0 ]; then
+  echo "Deploy failed"
+  exit ${code}
+fi
```
This looks like the error codes for the `scp` commands get ignored, and only a failure of the `ssh` command counts as a failure.

cori replied:
Good catch, thanks!

in https://github.com/FogCreek/Glitch-Community/blob/48dbb2095542e89237ceba7907fde9823ab4bb31/ci/publish-build-asset.sh#L36 I'm summing exit codes just to get success/failure, and the same would work here, although in means the exit codes themselves are useless (although as long at `set -x` is active you can see each step's exit code in the logs). Do you think that's a useful way to handle this, or do you know of a better one?

FWIW I pushed the change I mentioned above to test against the deploy.

eestolano replied:
Huh, I hadn't seen that technique before. That technique actually proceeds and runs subsequent commands if a command fails, and only fails out in the end. I think you're better off failing immediately -- both here and in `publish-build-asset.sh`. Otherwise, you could have some weird behavior -- for instance, if `scp`ing the deploy script to a box fails,, but the subsequent SSH command succeeds, you'll end up running a deploy on that box with a previous version of the deploy script.

cori replied:
updated!

eestolano commented on ci/env:
```diff
@@ -0,0 +1,42 @@
+# If we have CIRCLE_BRANCH this is running in CircleCI; set ENVIRONMENT appropriately
+# and use the CircleCI env for secrets
+if [[ -n "${CIRCLE_BRANCH}" ]];
+then
```
Is this `CIRCLE_BRANCH` check the only case where you expect to read an unset variable? You can test for its existence in a `set -u`-compatible way by replacing `if [[ -n "${CIRCLE_BRANCH}" ]];` with `if [[ -v CIRCLE_BRANCH ]];`

cori replied:
Ah, great tip! Thanks! Pushed!

eestolano commented on ci/local-deploy.sh:
```diff
@@ -0,0 +1,48 @@
+# THIS SCRIPT RUNS ON ALL COMMUNITY WORKERS
+# It manages the running site, gets the requested asset, and deploys and installs it. 
+# This script suffers from a bootstrap problem - if you change this script 
+# a simple deploy will probably not suffice and may require some special management
```
(Now that the bootstrap problem is resolved, this comment can go away, right?)

cori replied:
fixed

eestolano commented on ci/publish-build-asset.sh:
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env bash
+set -x  #   no -e or -o pipefail; we need to handle the exit code of the s3api call
```
If you put the `s3api` call in an `if` statement, then a failure won't make it exit with -e

cori replied:
I was resisting this because that syntax feels awkward to me, but better to _have_ the `-e`.

cori replied:
oh, and I think resolved

eestolano commented on ci/publish-build-asset.sh:
```diff
@@ -0,0 +1,47 @@
+    #   no file in s3
+    aws s3 cp --quiet "/home/circleci/$CIRCLE_SHA.tar.gz" "s3://$COMMUNITY_BOOTSTRAP_BUCKET"; code=$?
+    echo "$CIRCLE_SHA" > /home/circleci/LAST_DEPLOYED_SHA
+    aws s3 cp --quiet /home/circleci/LAST_DEPLOYED_SHA "s3://$COMMUNITY_BOOTSTRAP_BUCKET"; code=$(($code+$?)) # sum exit codes for a final success value
```
I imagine you don't actually want to update `LAST_DEPLOYED_SHA` if you failed to upload the tarball?

cori replied:
nope, thanks! fixed.

eestolano commented on ci/deploy.sh:
```diff
@@ -0,0 +1,51 @@
+
+  # try to avoid potential bootstrap problems with updates to the local deploy by uploading it 
+  # (and the corresponding env file) to /tmp and executing from there.
+  scp /home/circleci/project/ci/local-deploy.sh /home/circleci/project/ci/env "deploy@${name}.$ENVIRONMENT":/tmp; code=$?
```
whoops, missed this earlier. If this fails, the script will simply exit because of the `set -e`. Not the worst thing, since you're exiting anyway, but you miss out on the 'Deploy command failed' message.

Some options:
* stick it in an `if` statement
* `set +e` before the command, then `set -e` afterwards
